File: september/week1/RepeatedSubstringPattern.cpp
#include <iostream>
#include <algorithm>
#include <cstdio>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
	bool repeatedSubstringPattern(string s) {

		for (int i = 1; i < s.size(); i++) {
			string pivot = "";
			bool flag = true;
			if (s.size() % i != 0)
				continue;
			for (int j = 0; j < i; j++)
				pivot += s[j];

			for (int j = i; j < s.size(); j += i) {
				string cp = "";
				for (int k = j; k < j + i; k++) {
					cp += s[k];
				}

				if (pivot != cp) {
					flag = false;
					break;
				}
			}

			if (flag)
				return true;
		}
		return false;
	}
};

int main(void)
{
	Solution sol;
	cout << sol.repeatedSubstringPattern("abab") << endl;
	cout << sol.repeatedSubstringPattern("aba") << endl;
	cout << sol.repeatedSubstringPattern("abcabcabcabc") << endl;
	return 0;
}

File: september/week1/repeated_substring_pattern.py
class Solution:
    def repeatedSubstringPattern(self, s: str) -> bool:
        n = len(s)
        dp = [0]*n
        i=1;j=0
        while i<n:
            while j>0 and s[i] != s[j]:
                j = dp[j-1]
            if s[i] == s[j]:
                j += 1
                dp[i] += j
            i += 1
        cutn = n-dp[n-1]
        tar = s[:cutn]
        if cutn==n or n%cutn:
            return False
        else:
            if tar*(n//cutn) == s:
                return True
            else:
                return False

if __name__ == "__main__":
    sol = Solution()
    print(sol.repeatedSubstringPattern("abab"))
    print(sol.repeatedSubstringPattern("aba"))
    print(sol.repeatedSubstringPattern("abcabcabcabc"))
    print(sol.repeatedSubstringPattern("abac"))
